txnsync/peer.go

```diff
@@ -497,7 +504,44 @@ func incomingPeersOnly(peers []*Peer) (incomingPeers []*Peer) {
 
 // incoming related functions
 
-func (p *Peer) addIncomingBloomFilter(round basics.Round, incomingFilter *testableBloomFilter, currentRound basics.Round) {
+// addIncomingBloomFilter keeps the most recent {maxIncomingBloomFilterHistory} filters
+func (p *Peer) addIncomingBloomFilter(round basics.Round, incomingFilter *testableBloomFilter, currentRound basics.Round, maxIncomingFilterElements int) {
+	bf := incomingBloomFilter{
+		round:  round,
+		filter: incomingFilter,
+	}
+	activeElements := int(bf.filter.elementsFiltered)
+	if len(p.recentIncomingBloomFilters) < maxIncomingBloomFilterHistory {
+		p.recentIncomingBloomFilters = append(p.recentIncomingBloomFilters, bf)
+		prev := len(p.recentIncomingBloomFilters) - 2
+		for prev >= 0 {
+			if activeElements >= maxIncomingFilterElements {
+				p.recentIncomingBloomFilters[prev].filter = nil
+			} else {
+				activeElements += int(p.recentIncomingBloomFilters[prev].filter.elementsFiltered)
+			}
+			prev--
+		}
+		return
+	}
+
+	p.recentIncomingBloomFilters[p.recentIncomingBloomFiltersNext] = bf
+	prev := (p.recentIncomingBloomFiltersNext + maxIncomingBloomFilterHistory - 1) % maxIncomingBloomFilterHistory
+	for prev != p.recentIncomingBloomFiltersNext {
+		if activeElements >= maxIncomingFilterElements {
+			p.recentIncomingBloomFilters[prev].filter = nil
+		} else {
+			activeElements += int(p.recentIncomingBloomFilters[prev].filter.elementsFiltered)
+		}
+		prev = (prev + maxIncomingBloomFilterHistory - 1) % maxIncomingBloomFilterHistory
+	}
+
+	p.recentIncomingBloomFiltersNext = (p.recentIncomingBloomFiltersNext + 1) % maxIncomingBloomFilterHistory
+
+	// TODO: limit total elements to maxIncomingFilterElements
+}
```

Comment:
you would still need
```golang
// reset the counter, since we might need to re-evaluate some of the transaction group with the new bloom filter.
	p.lastTransactionSelectionTracker.roll(incomingFilter.encodingParams.Offset, incomingFilter.encodingParams.Modulator)
```